postgres_exporter/common/setup_metric_views.sql

```diff
@@ -0,0 +1,174 @@
+-- #########################
+-- Metric View Objects
+-- #########################
+
+CREATE SCHEMA IF NOT EXISTS monitor;
+
+-- Preserve existing data if users added more views or changed schedule. Allows setup call to be idempodent.
+DO $pgmonitor$
+DECLARE
+v_exists    smallint;
+BEGIN
+
+    SELECT count(*) INTO v_exists FROM information_schema.tables WHERE table_schema = 'monitor' AND table_name = 'metric_views';
+    IF v_exists > 0 THEN
+        CREATE TEMPORARY TABLE metric_views_preserve_temp (LIKE monitor.metric_views);
+        INSERT INTO metric_views_preserve_temp SELECT * FROM monitor.metric_views;
+        DROP TABLE monitor.metric_views;
+    END IF;
+
+    CREATE TABLE IF NOT EXISTS monitor.metric_views (
+        view_schema text NOT NULL DEFAULT 'monitor'
+        , view_name text NOT NULL
+        , concurrent_refresh boolean NOT NULL DEFAULT true
+        , run_interval interval NOT NULL
+        , last_run timestamptz
+        , active boolean NOT NULL DEFAULT true
+        , scope text NOT NULL default 'global'
+        , CONSTRAINT metric_views_pk PRIMARY KEY (view_schema, view_name)
+        , CONSTRAINT metric_views_scope_ck CHECK (scope IN ('global', 'database'))
+    );
+
+    IF v_exists > 0 THEN
+        INSERT INTO monitor.metric_views SELECT * FROM metric_views_preserve_temp;
+        DROP TABLE metric_views_preserve_temp;
+    END IF;
+
+END
+$pgmonitor$;
+
+
+DROP PROCEDURE IF EXISTS monitor.refresh_metric_views (text, text);
+CREATE PROCEDURE monitor.refresh_metric_views (p_view_schema text DEFAULT 'monitor', p_view_name text DEFAULT NULL)
+    LANGUAGE plpgsql
+    AS $$
+DECLARE
+
+v_loop_sql           text;
+v_refresh_sql        text;
+v_recovery           boolean;
+v_row                record;
+
+BEGIN
+
+SELECT pg_is_in_recovery() INTO v_recovery;
+IF v_recovery THEN
+    RAISE DEBUG 'Database instance in recovery mode. Exiting without view refresh';
+    RETURN;
+END IF;
+
+v_loop_sql := format('SELECT view_schema, view_name, concurrent_refresh, run_interval, last_run 
+                        FROM monitor.metric_views
+                        WHERE active');
+
+IF p_view_name IS NOT NULL THEN
+    v_loop_sql := format('%s AND view_schema = %L AND view_name = %L', v_loop_sql, p_view_schema, p_view_name);
+END IF;
+
+FOR v_row IN EXECUTE v_loop_sql LOOP
+
+    IF ((CURRENT_TIMESTAMP - v_row.last_run) > v_row.run_interval) OR (v_row.last_run IS NULL) THEN
```

Comment:
Instead of having this IF statement, suggest moving the timestamp check to the SELECT statements WHERE clause.

Reply:
I think that would make the view itself be a little bit more complicated than necessary. This function is still needed to determine primary/replica and to update the last_run value, so just going to leave the check in here for now.

Reply:
Oh nevermind, you mean the select statement to pull the list of views back. Yeah, that makes sense. Sorry.

Reply:
Done. Thanks for the suggestion!